api/python/quilt3/data_transfer.py

```diff
@@ -530,43 +530,61 @@ def upload_part(i, start, end):
+
+
+def _reuse_remote_file(ctx: WorkerContext, size: int, src_path: str, dest_bucket: str, dest_path: str):
     # Optimization: check if the remote file already exists and has the right ETag,
     # and skip the upload.
-    if size >= UPLOAD_ETAG_OPTIMIZATION_THRESHOLD:
-        try:
-            params = dict(Bucket=dest_bucket, Key=dest_path)
-            s3_client = ctx.s3_client_provider.find_correct_client(S3Api.HEAD_OBJECT, dest_bucket, params)
-            resp = s3_client.head_object(**params, ChecksumMode='ENABLED')
-        except ClientError:
-            # Destination doesn't exist, so fall through to the normal upload.
-            pass
-        except S3NoValidClientError:
-            # S3ClientProvider can't currently distinguish between a user that has PUT but not LIST permissions and a
-            # user that has no permissions. If we can't find a valid client, proceed to the upload stage anyway.
-            pass
-        else:
-            # Check the ETag.
-            dest_size = resp['ContentLength']
-            dest_etag = resp['ETag']
-            dest_version_id = resp.get('VersionId')
-            if size == dest_size and resp.get('ServerSideEncryption') != 'aws:kms':
-                src_etag = _calculate_etag(src_path)
-                if src_etag == dest_etag:
-                    # Nothing more to do. We should not attempt to copy the object because
-                    # that would cause the "copy object to itself" error.
-                    # TODO: Check SHA256 before checking ETag?
-                    s3_checksum = resp.get('ChecksumSHA256')
-                    if s3_checksum is None:
-                        checksum = None
-                    elif '-' in s3_checksum:
-                        checksum, _ = s3_checksum.split('-', 1)
-                    else:
-                        checksum = _simple_s3_to_quilt_checksum(s3_checksum)
-                    ctx.progress(size)
-                    ctx.done(PhysicalKey(dest_bucket, dest_path, dest_version_id), checksum)
-                    return  # Optimization succeeded.
+    if size < UPLOAD_ETAG_OPTIMIZATION_THRESHOLD:
+        return None
+    try:
+        params = dict(Bucket=dest_bucket, Key=dest_path)
+        s3_client = ctx.s3_client_provider.find_correct_client(S3Api.HEAD_OBJECT, dest_bucket, params)
+        resp = s3_client.head_object(**params, ChecksumMode="ENABLED")
+    except ClientError:
+        # Destination doesn't exist, so fall through to the normal upload.
+        pass
+    except S3NoValidClientError:
+        # S3ClientProvider can't currently distinguish between a user that has PUT but not LIST permissions and a
+        # user that has no permissions. If we can't find a valid client, proceed to the upload stage anyway.
+        pass
+    else:
+        dest_size = resp["ContentLength"]
+        if dest_size != size:
+            return None
+        # XXX: shouldn't we check part sizes?
```

Comment:
i'd say we should, or at least check the number of parts is correct

Reply:
now it checks number of parts